rfcs/2022-12-09-541-reload-api.md

```diff
@@ -0,0 +1,208 @@
+process is included with the RFC PR to help show dependencies. The next step
+would likely be to extract a `TopologyController` or similar struct that owns
+the topology and everything that needs to change along with it. That struct can
+then present a simple API for reloads that ensures everything is done in a
+consistent matter, no matter the source.
+
+Once that struct is in place, it can be wrapped in an `Arc<Mutex<_>>` or
+similar such that it can be shared across both the existing `Signal`-focused
+run loop and a new control server. The new control server itself can be a very
+simple task spawned prior to entering the main loop. As a local-only control
+API, it can different from most HTTP servers in a few ways:
+
+- By listening on a Unix socket instead of the network, it can rely on
+  filesystem permissions for authentication rather than its own scheme.
+- Because it's designed for use with a single local client, it does not need to
+  support concurrent processing of any kind. If it can't immediately acquire
+  the lock on the `TopologyController`, it can fail and tell the client to try
+  again.
+
+With the server and controller in place, the path for synchronous feedback will
+exist. The next step is to adjust the implementation of
+`RunningTopology::reload_config_and_respawn` away from the current return value
+of `Result<bool, ()>` and towards something that actually returns error
+messages. This will involve a fair bit of work down into the reloading stack,
+where we currently tend to emit errors as they happen. The likely solution will
+be refactoring to build up an error report of some kind that can be reported
+and/or returned at the end of the whole process.
+
+## Rationale
+
+This approach is designed to be relatively minimal while still achieving all of
+the functionality needed for a good user experience. There are certainly deeper
+changes that could (and should) be made around the design of this part of
+Vector, and what we've set out here should be a good first step towards that
+future while providing immediate user value. This design should also be
+relatively easy to expand and evolve over time should be decide to adjust any
+aspect of it.
+
+## Drawbacks
+
+The main drawback is the overlap in functionality with existing pieces of
+Vector. These include:
+
+- The GraphQL API: This is already "the API" and some confusion could result
+  from introducing another (see alternatives section below for a discussion of
+  why we're not reusing it).
+- Providers: This is an existing path for configurable ways to provide
+  configuration outside of the standard config file. Some but not all of the
+  functionality here could be implemented as a provider.
+- File-based configs: It could be confusing that this approach is not exclusive
+  to the normal config loading/reloading processes being active. In theory,
+  some strange interactions between the two are possible.
+
+## Prior Art
```

Comment:
I would be curious to see some comparisons with other tools that support configuration reloading (like haproxy?).

Reply:
[Envoy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/operations/hot_restart), [Consul](https://developer.hashicorp.com/consul/commands/reload) (likely all of the HashiSuite) are another two examples off the top of my head.

Reply:
Meant to fill this out, oops. Just pushed an update.